sensor/FamilyProtectPlugin.js

```diff
@@ -0,0 +1,193 @@
+Promise.promisifyAll(fs);
+
+const DNSMASQ = require('../extension/dnsmasq/dnsmasq.js');
+const dnsmasq = new DNSMASQ();
+
+const exec = require('child-process-promise').exec;
+
+const fc = require('../net2/config.js');
+
+class FamilyProtectPlugin extends Sensor {
+    async run() {
+        this.systemSwitch = false;
+        this.adminSystemSwitch = false;
+        this.enabledMacAddresses = {};
+        extensionManager.registerExtension("family", this, {
+            applyPolicy: this.applyPolicy,
+            start: this.start,
+            stop: this.stop
+        });
+        await exec(`mkdir -p ${dnsmasqConfigFolder}`);
+        if (fc.isFeatureOn("family_protect")) {
+            await this.globalOn();
+        } else {
+            await this.globalOff();
+        }
+        fc.onFeature("family_protect", async (feature, status) => {
+            if (feature !== "family_protect") {
+                return;
+            }
+            if (status) {
+                await this.globalOn();
+            } else {
+                await this.globalOff();
+            }
+        })
+
+        await this.job();
+        this.timer = setInterval(async () => {
+            return this.job();
+        }, this.config.refreshInterval || 3600 * 1000); // one hour by default
+    }
+
+    async job() {
+        await this.applyFamilyProtect();
+    }
+
+    async apiRun() {
+
+    }
+
+    async applyPolicy(host, ip, policy) {
+        log.info("Applying family protect policy:", ip, policy);
+        try {
+            if (ip === '0.0.0.0') {
+                if (policy == true) {
+                    this.systemSwitch = true;
+                } else {
+                    this.systemSwitch = false;
+                }
+                return this.applySystemFamilyProtect();
+            } else {
+                const macAddress = host && host.o && host.o.mac;
+                if (macAddress) {
+                    if (policy == true) {
+                        this.enabledMacAddresses[macAddress] = 1;
+                    } else {
+                        delete this.enabledMacAddresses[macAddress];
+                    }
+                    return this.applyDeviceFamilyProtect(macAddress);
+                }
+            }
+        } catch (err) {
+            log.error("Got error when applying family protect policy", err);
+        }
+    }
+
+    async applyFamilyProtect() {
+        await this.applySystemFamilyProtect();
+        for (const macAddress in this.enabledMacAddresses) {
+            await this.applyDeviceFamilyProtect(macAddress);
+        }
+    }
+
+    async applySystemFamilyProtect() {
+        this.familyDnsAddr((err, dnsaddrs) => {
+            if (this.systemSwitch && this.adminSystemSwitch) {
+                return this.systemStart(dnsaddrs);
+            } else {
+                return this.systemStop(dnsaddrs);
+            }
+        });
+    }
+
+    async applyDeviceFamilyProtect(macAddress) {
```

Comment:
applyDeviceFamilyProtect should also use 'adminSystemSwitch' to determine if changes should be applied. If feature is off, neither system nor device family protect should be applied.

Reply:
this could also be a defect of SafeSearch feature?

they have same code structure